spec/type_references_and_ownership.md

```diff
@@ -0,0 +1,331 @@
+    Subscription string
+
+    // The resource group the Azure resource is in. If empty, this defaults to the resource group the resource 
+    // holding the reference is in.
+    ResourceGroup string
+
+    // The name of the resource. This is a subset of a fully qualified ARM ID - if you're 
+    // referencing a top level resource (for example a virtual network): Microsoft.Network/virtualNetworks/{network}
+    // If you're referencing a subnet: Microsoft.Network/virtualNetworks/{network}/subnets/{subnet}
+    // TODO: There are a ton of options for how we do this... some of the requiring more manual work than others... we should discuss this
+    Name string
+}
+```
+
+This approach seems to have the most flexibility and power. It allows easy reference to static Azure resources
+not managed by Kubernetes but also allows users to stay entirely within the Kubernetes ecosystem and refer to everything by
+its Kubernetes name.
+
+The primary downside of this approach is complexity - references aren't just a single string they're a complex
+object and the customer will need to choose which flavor they want each time they use a reference.
+
+#### Implementation
+There are two key implementation details for this approach:
+1. We must identify all resource references in the JSON schema and transform their type to `AugmentedKnownTypeReference`.
+   There is no specific marker which means: "This field is a reference" - most are called `id` but that's not a guarantee.
+   For example on the [VirtualMachineScaleSetIPConfigurationProperties](https://schema.management.azure.com/schemas/2019-07-01/Microsoft.Compute.json) 
+   the `subnet` field is of custom type `ApiEntityReference`, which has an `id` field where you put the ARM ID for a subnet.
+   This may require some manual work. One thing we can investigate doing long term
+   is see if there's a way to get teams to annotate "links" in their Swagger somehow...
+2. We have to be able to build the fully qualified ARM ID for any referenced resource.
+
+### How to represent ownership and dependent resources
+Many resources have owners but do not have an `id` field referring to their owner. Similarly, many resources with dependent resources may not
+have a field which refers to them.
+**We need to code generate a field which allows users of the CRDs to express ownership**.
+As discussed in [what ARM does](#What-ARM-does), this is the `resources` property in the ARM deployment templates.
+
+When we determine that a resource is a dependent resource of another resource kind, we will code-generate
+an `OwnerRef` (actual name TBD) property in the dependent resource `Spec`. This will also include an annotation 
+about the expected type of the resource.
+
+```go
+type SubnetSpec struct {
+    OwnerRef AugmentedKnownTypeReference `json:"ownerRef" group:"microsoft.network.infra.azure.com" kind:"VirtualNetwork"`
```

Comment:
Why would `OwnerRef` exist if the `Subnet` resource has an `subnetResource.OwnershipReferences` contains a `corev1.OwnershipReference` corresponding to the VNET? This field seems superfluous.

Reply:
@devigned and I talked about this offline.

This is basically inverting the ownership approach from what k8s-infra is doing today. Instead of the parent including references to its dependents, the dependent resource includes a reference to its parent. This is in-line with what Azure is actually doing in the ARM templates too (where 

The takeaways from the discussion were:
- The name `OwnerRef` probably needs to be reconsidered. It's confusing/collides with `OwnerReferences` in the Kubernetes object metadata.
- The reason that the `KnownTypeReference` in k8s-infra went from parent -> child is because there are a number of resources where the child resource _is a property_ of the parent. A good example of this is `RouteTable`. David and I went through `RouteTable` as an exercise though and confirmed that you can create a route either by putting it side-by-side and using `dependsOn` in the ARM template, OR by putting the resource embedded in the parent resource. Both are actually identical in terms of what they do in ARM. Not every resource has this kind of relationship though (for example `BatchAccount` -> `Pool` doesn't) so we can't rely on that. After much discussion it was agreed that dependent resources referencing their owners (as they do in ARM templates today when deployed side-by-side with `dependsOn` seems a reasonable approach to proceed with).

The other key highlight is that, this field (called `OwnerRef` here, final name TBD) is what we will use in the controller as metadata to set the actual `ownershipReferences` property in `k8s` metadata.